src/segmentation.rs

```diff
@@ -31,99 +40,226 @@ pub fn segmentation_optimize<T: Pixel>(
+      *delta *= 0.8;
+    }
+  });
+
+  /* Get all unique values in the intentionally unsorted array (its a LUT) */
+  let mut uniq_array = [0usize; MAX_SEGMENTS];
+  let mut num_segments = 0;
+  for i in 0..MAX_SEGMENTS {
+    let mut seen_match = false;
+    for j in 0..num_segments {
+      if remap_segment_tab[i] == uniq_array[j] {
+        seen_match = true;
+      }
+    }
+    if !seen_match {
+      uniq_array[num_segments] = remap_segment_tab[i];
+      num_segments += 1;
+    }
+  }
+
+  let mut seg_delta = [0f64; MAX_SEGMENTS];
+  for i in 0..num_segments {
+    /* Collect all used segment deltas into the actual segment map */
+    seg_delta[i] = tmp_delta[uniq_array[i]];
+
+    /* Remap the LUT to make it match the layout of the seg deltaq map */
+    for j in 0..MAX_SEGMENTS {
+      if remap_segment_tab[j] == uniq_array[i] {
+        remap_segment_tab[j] = i;
+      }
+    }
+  }
+
+  fs.segmentation.segment_map = remap_segment_tab;
+
+  for i in 0..num_segments {
+    fs.segmentation.features[i][SegLvl::SEG_LVL_ALT_Q as usize] = true;
+    fs.segmentation.data[i][SegLvl::SEG_LVL_ALT_Q as usize] =
+      (seg_delta[i].round() as i16).max(offset_lower_limit);
+  }
+}
+
 pub fn select_segment<T: Pixel>(
   fi: &FrameInvariants<T>, ts: &TileStateMut<'_, T>, tile_bo: TileBlockOffset,
   bsize: BlockSize, skip: bool,
 ) -> std::ops::RangeInclusive<u8> {
-  use crate::api::SegmentationLevel;
-  use crate::rdo::spatiotemporal_scale;
-  use arrayvec::ArrayVec;
-
   // If skip is true or segmentation is turned off, sidx is not coded.
   if skip || !fi.enable_segmentation {
     return 0..=0;
   }
 
-  let segment_2_is_lossless = fi.base_q_idx as i16
-    + ts.segmentation.data[2][SegLvl::SEG_LVL_ALT_Q as usize]
-    < 1;
-
-  if fi.config.speed_settings.segmentation == SegmentationLevel::Full {
-    return if segment_2_is_lossless { 0..=1 } else { 0..=2 };
-  }
-
   let frame_bo = ts.to_frame_block_offset(tile_bo);
   let scale = spatiotemporal_scale(fi, frame_bo, bsize);
 
-  // TODO: Replace this calculation with precomputed scale thresholds.
-  let seg_ac_q: ArrayVec<_, 3> = if fi.enable_segmentation {
-    use crate::quantize::ac_q;
-    (0..=2)
-      .map(|sidx| {
-        ac_q(
-          (fi.base_q_idx as i16
-            + ts.segmentation.data[sidx][SegLvl::SEG_LVL_ALT_Q as usize])
-            .max(0)
-            .min(255) as u8,
-          0,
-          fi.sequence.bit_depth,
-        )
-      })
-      .collect()
-  } else {
-    Default::default()
-  };
-
-  let sidx = if scale.mul_u64(seg_ac_q[1] as u64) < seg_ac_q[0] as u64 {
-    1
-  } else if !segment_2_is_lossless
-    && scale.mul_u64(seg_ac_q[2] as u64) > seg_ac_q[0] as u64
-  {
-    2
-  } else {
-    0
-  };
-
-  sidx..=sidx
+  // SAFETY: We know that every segment is between 0-7
+  unsafe {
+    let sidx = *ts
+      .segmentation
+      .segment_map
+      .get_unchecked(segment_idx_from_distortion(scale) as usize)
```

Comment:
will indexing with & 0x7 elide the bounds check here?

Reply:
It seems that, yes, it does. That's a neat trick I'll have to keep in mind from now on.

I'm going to code it as a modulus because I think it makes the intention clearer, but according to Godbolt `%` and `&` generate the same asm: https://rust.godbolt.org/z/EnevWG6GG